AZ3166/src/libraries/WebSocket/src/WebSocketClient.cpp

```diff
@@ -0,0 +1,475 @@
+
+WebSocketReceiveResult *WebSocketClient::receive(char *message)
+{
+    if (_tcpSocket == NULL)
+    {
+        Serial.println("ERROR: unable to receive data when WebSocket is disconnected.");
+        return NULL;
+    }
+
+    int i = 0;
+    uint32_t payloadLength;
+    char recvByte = 0;
+    char opcode = 0;
+    char c;
+    char mask[4] = {0, 0, 0, 0};
+    bool isMasked = false;
+    bool isFinal = false;
+    Timer timer;
+
+    // Read opcode
+    timer.start();
+    while (true)
+    {
+        if (timer.read() > 10)
+        {
+            //Serial.printf("ERROR: timeout ws, timer: %d\r\n", timer.read());
+            return NULL;
+        }
+
+        int res = _tcpSocket->recv(&recvByte, 1);
+
+        if (res == 1)
+        {
+            _tcpSocket->set_timeout(2000);
+            opcode = recvByte & 0x7F;
+
+            // opcode for data frames
+            if (opcode == WS_OPCODE_CONT || opcode == WS_OPCODE_TEXT || opcode == WS_OPCODE_BINARY)
+            {
+                if (opcode == WS_OPCODE_TEXT)
+                {
+                    messageType = WS_Message_Text;
+                }
+                else if (opcode == WS_OPCODE_BINARY)
+                {
+                    messageType = WS_Message_Binary;
+                }
+
+                isFinal = ((recvByte & 0x80) == 0x80);
+                break;
+            }
+            // opcode for connection close
+            else if (opcode == WS_OPCODE_CLOSE)
+            {
+                close();
+            }
+        }
+        else if (res < 0 && res != NSAPI_ERROR_WOULD_BLOCK)
+        {
+            Serial.printf("ERROR: Socket receive failed, res: %d, opcode: %x\r\n", res, opcode);
+            return NULL;
+        }
+    }
+
+    // Parse payload length
+    readChar(&c);
+    payloadLength = c & 0x7f;
+    isMasked = c & 0x80;
+    if (payloadLength == 126)
+    {
+        readChar(&c);
+        payloadLength = c << 8;
+        readChar(&c);
+        payloadLength += c;
+    }
+    else if (payloadLength == 127)
+    {
+        payloadLength = 0;
+        for (int i = 0; i < 8; i++)
+        {
+            readChar(&c);
+            payloadLength += (c << (7 - i) * 8);
+        }
+    }
+
+    if (payloadLength == 0)
+    {
+        return NULL;
+    }
+
+    if (isMasked)
+    {
+        for (i = 0; i < 4; i++)
+        {
+            readChar(&c);
+            mask[i] = c;
+        }
+    }
+
+    int nb = read(message, payloadLength, payloadLength);
+    if (nb != payloadLength)
+        return NULL;
+
+    for (i = 0; i < payloadLength; i++)
+    {
+        message[i] = message[i] ^ mask[i % 4];
+    }
+
+    message[payloadLength] = '\0';
+
+    WebSocketReceiveResult *receiveResult = new WebSocketReceiveResult();
+    receiveResult->isEndOfMessage = isFinal;
+    receiveResult->length = payloadLength;
+    receiveResult->messageType = messageType;
+
+    return receiveResult;
+}
+
+bool WebSocketClient::close()
+{
+    if (_tcpSocket == NULL)

```

Comment:
As I remembered, a close frame needed to be sent before close. Or extra operation will needed on server side. Add a todo here ?